op-challenger/fault/monitor.go

```diff
@@ -2,63 +2,94 @@ package fault
+type gameCreator func(address common.Address) (gameAgent, error)
+type blockNumberFetcher func(ctx context.Context) (uint64, error)
+
+// gameSource loads information about the games available to play
+type gameSource interface {
+	FetchAllGamesAtBlock(ctx context.Context, blockNumber *big.Int) ([]FaultDisputeGame, error)
 }
 
-func MonitorGame(ctx context.Context, logger log.Logger, agreeWithProposedOutput bool, actor Actor, caller GameInfo) error {
-	logger.Info("Monitoring fault dispute game", "agreeWithOutput", agreeWithProposedOutput)
+type gameMonitor struct {
+	logger           log.Logger
+	source           gameSource
+	createGame       gameCreator
+	fetchBlockNumber blockNumberFetcher
+	allowedGame      common.Address
+	games            map[common.Address]gameAgent
+}
 
-	for {
-		done := progressGame(ctx, logger, agreeWithProposedOutput, actor, caller)
-		if done {
-			return nil
+func newGameMonitor(logger log.Logger, fetchBlockNumber blockNumberFetcher, allowedGame common.Address, source gameSource, createGame gameCreator) *gameMonitor {
+	return &gameMonitor{
+		logger:           logger,
+		source:           source,
+		createGame:       createGame,
+		fetchBlockNumber: fetchBlockNumber,
+		allowedGame:      allowedGame,
+		games:            make(map[common.Address]gameAgent),
+	}
+}
+
+func (m *gameMonitor) progressGames(ctx context.Context) error {
+	blockNum, err := m.fetchBlockNumber(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to load current block number: %w", err)
+	}
+	games, err := m.source.FetchAllGamesAtBlock(ctx, new(big.Int).SetUint64(blockNum))
+	if err != nil {
+		return fmt.Errorf("failed to load games: %w", err)
+	}
+	for _, game := range games {
+		if m.allowedGame != (common.Address{}) && m.allowedGame != game.Proxy {
+			m.logger.Debug("Skipping game not on allow list", "game", game.Proxy)
+			continue
 		}
-		select {
-		case <-time.After(300 * time.Millisecond):
-		// Continue
-		case <-ctx.Done():
-			return ctx.Err()
+		if err := m.progressGame(ctx, game); err != nil {
+			m.logger.Error("Error while progressing game", "game", game.Proxy, "err", err)
 		}
 	}
+	return nil
 }
 
-// progressGame checks the current state of the game, and attempts to progress it by performing moves, steps or resolving
-// Returns true if the game is complete or false if it needs to be monitored further
-func progressGame(ctx context.Context, logger log.Logger, agreeWithProposedOutput bool, actor Actor, caller GameInfo) bool {
-	logger.Trace("Checking if actions are required")
-	if err := actor.Act(ctx); err != nil {
-		logger.Error("Error when acting on game", "err", err)
+func (m *gameMonitor) progressGame(ctx context.Context, gameData FaultDisputeGame) error {
+	game, ok := m.games[gameData.Proxy]
+	if !ok {
+		newGame, err := m.createGame(gameData.Proxy)
+		if err != nil {
+			return fmt.Errorf("failed to progress game %v: %w", gameData.Proxy, err)
+		}
+		m.games[gameData.Proxy] = newGame
+		game = newGame
 	}
```

Comment:
todo(refcell): stack ontop of this pr: refactor out and unit test a new `FetchOrCreateGame()` function

Reply:
I've done this refactoring. Once you pull out the fetch or create logic, the `progressGame` method isn't worth having so I've inlined it into `progressGames`.